packages/pulseaudio/module-sles-sink.c

```diff
@@ -125,23 +130,197 @@ static void sles_callback(SLBufferQueueItf bqPlayerBufferQueue, void *userdata)
+        case PA_CHANNEL_POSITION_TOP_FRONT_CENTER:      return SL_SPEAKER_TOP_FRONT_CENTER;
+        case PA_CHANNEL_POSITION_TOP_REAR_LEFT:         return SL_SPEAKER_TOP_BACK_LEFT;
+        case PA_CHANNEL_POSITION_TOP_REAR_RIGHT:        return SL_SPEAKER_TOP_BACK_RIGHT;
+        case PA_CHANNEL_POSITION_TOP_REAR_CENTER:       return SL_SPEAKER_TOP_BACK_CENTER;
+        case PA_CHANNEL_POSITION_INVALID:               return (SLuint32)(-1);
+        default:                                        return (SLuint32)(-1);
+    }
+}
+
+// sl_speaker_to_pa_channel_position converts an OpenSL ES speaker to the
+// equivalent PulseAudio channel position. If the channel does not have a PA
+// equivalent or is invalid, PA_CHANNEL_POSITION_INVALID is returned. Note that
+// this function does not hand the edge case where (SL_SPEAKER_FRONT_CENTER &&
+// channels == 1) is PA_CHANNEL_POSITION_MONO.
+//
+// https://freedesktop.org/software/pulseaudio/doxygen/channelmap_8h.html#af1cbe2738487c74f99e613779bd34bf2
+// https://www.khronos.org/files/opensl_es_1_0_provisional_specs.pdf#9.2.47
+static enum pa_channel_position sl_speaker_to_pa_channel_position(SLuint32 x) {
+    switch (x) {
+        //case SL_SPEAKER_FRONT_CENTER:        return PA_CHANNEL_POSITION_MONO;
+        case SL_SPEAKER_FRONT_LEFT:            return PA_CHANNEL_POSITION_FRONT_LEFT;
+        case SL_SPEAKER_FRONT_RIGHT:           return PA_CHANNEL_POSITION_FRONT_RIGHT;
+        case SL_SPEAKER_FRONT_CENTER:          return PA_CHANNEL_POSITION_FRONT_CENTER;
+        case SL_SPEAKER_BACK_CENTER:           return PA_CHANNEL_POSITION_REAR_CENTER;
+        case SL_SPEAKER_BACK_LEFT:             return PA_CHANNEL_POSITION_REAR_LEFT;
+        case SL_SPEAKER_BACK_RIGHT:            return PA_CHANNEL_POSITION_REAR_RIGHT;
+        case SL_SPEAKER_LOW_FREQUENCY:         return PA_CHANNEL_POSITION_LFE;
+        case SL_SPEAKER_FRONT_LEFT_OF_CENTER:  return PA_CHANNEL_POSITION_FRONT_LEFT_OF_CENTER;
+        case SL_SPEAKER_FRONT_RIGHT_OF_CENTER: return PA_CHANNEL_POSITION_FRONT_RIGHT_OF_CENTER;
+        case SL_SPEAKER_SIDE_LEFT:             return PA_CHANNEL_POSITION_SIDE_LEFT;
+        case SL_SPEAKER_SIDE_RIGHT:            return PA_CHANNEL_POSITION_SIDE_RIGHT;
+        case SL_SPEAKER_TOP_CENTER:            return PA_CHANNEL_POSITION_TOP_CENTER;
+        case SL_SPEAKER_TOP_FRONT_LEFT:        return PA_CHANNEL_POSITION_TOP_FRONT_LEFT;
+        case SL_SPEAKER_TOP_FRONT_RIGHT:       return PA_CHANNEL_POSITION_TOP_FRONT_RIGHT;
+        case SL_SPEAKER_TOP_FRONT_CENTER:      return PA_CHANNEL_POSITION_TOP_FRONT_CENTER;
+        case SL_SPEAKER_TOP_BACK_LEFT:         return PA_CHANNEL_POSITION_TOP_REAR_LEFT;
+        case SL_SPEAKER_TOP_BACK_RIGHT:        return PA_CHANNEL_POSITION_TOP_REAR_RIGHT;
+        case SL_SPEAKER_TOP_BACK_CENTER:       return PA_CHANNEL_POSITION_TOP_REAR_CENTER;
+        default:                               return -1;
+    }
+}
+
```

Comment:
Do the above with a single array (of arrays of pairs) and a match loop as each of the functions?

Reply:
I'm not sure I see the advantage for that? If it's to have a single source of truth for the data, an X-macro would probably be be better so we can keep the efficiency, simplicity, and compile-time checking of a switch.

Reply:
Why repeat the same thing? Doesn't it only makes it harder to read / maintain? (Even one out of the three is already too long...)
If you have another approach to avoid that, it's fine as well. Or,  if you insist, splitting them into a header file is acceptable as well.

Reply:
Done.